File: src/models/ReceiveV1.ts
export type ReceiveV1Response = Array<ReceiveV1ResponseElement>;

export interface Envelope {
  source: string;
  sourceNumber: string;
  sourceUuid: string;
  sourceName: string;
  sourceDevice: number;
  timestamp: number;
}

export class SyncMessageEnvelope implements Envelope {
  source: string;
  sourceNumber: string;
  sourceUuid: string;
  sourceName: string;
  sourceDevice: number;
  timestamp: number;
  syncMessage: SyncMessage;

  constructor(
    source: string,
    sourceNumber: string,
    sourceUuid: string,
    sourceName: string,
    sourceDevice: number,
    timestamp: number,
    syncMessage: SyncMessage,
  ) {
    this.source = source;
    this.sourceNumber = sourceNumber;
    this.sourceUuid = sourceUuid;
    this.sourceName = sourceName;
    this.sourceDevice = sourceDevice;
    this.timestamp = timestamp;
    this.syncMessage = syncMessage;
  }

  static assertType(envelope: Envelope): envelope is SyncMessageEnvelope {
    return (envelope as SyncMessageEnvelope).syncMessage !== undefined;
  }
}

export interface SyncMessage {
  sentMessage: SentMessage;
}

export interface SentMessage {
  destination: null;
  destinationNumber: null;
  destinationUuid: null;
  timestamp: number;
  message: string;
  expiresInSeconds: number;
  viewOnce: boolean;
  groupInfo: string;
}

export class MessageEnvelope implements Envelope {
  source: string;
  sourceNumber: string;
  sourceUuid: string;
  sourceName: string;
  sourceDevice: number;
  timestamp: number;
  dataMessage: DataMessage;

  constructor(
    source: string,
    sourceNumber: string,
    sourceUuid: string,
    sourceName: string,
    sourceDevice: number,
    timestamp: number,
    dataMessage: DataMessage,
  ) {
    this.source = source;
    this.sourceNumber = sourceNumber;
    this.sourceUuid = sourceUuid;
    this.sourceName = sourceName;
    this.sourceDevice = sourceDevice;
    this.timestamp = timestamp;
    this.dataMessage = dataMessage;
  }

  static fromSyncMessageEnvelope(syncMessageEnvelope: SyncMessageEnvelope): MessageEnvelope {
    const syncMessage = syncMessageEnvelope.syncMessage;
    const sentMessage = syncMessage.sentMessage;
    const dataMessage = new DataMessage(
      sentMessage.timestamp,
      sentMessage.message,
      sentMessage.expiresInSeconds,
      sentMessage.viewOnce,
      sentMessage.groupInfo,
    );
    return new MessageEnvelope(
      syncMessageEnvelope.source,
      syncMessageEnvelope.sourceNumber,
      syncMessageEnvelope.sourceUuid,
      syncMessageEnvelope.sourceName,
      syncMessageEnvelope.sourceDevice,
      syncMessageEnvelope.timestamp,
      dataMessage,
    );
  }

  static assertType(envelope: Envelope): envelope is MessageEnvelope {
    return (envelope as MessageEnvelope).dataMessage !== undefined;
  }
}

export class DataMessage {
  constructor(
    public timestamp: number,
    public message: string,
    public expiresInSeconds: number,
    public viewOnce: boolean,
    public groupInfo: string,
  ) {}
}

export interface ReceiveV1ResponseElement {
  envelope: Envelope;
  account: string;
}

File: mock_server/src/models/receive_v1.rs
use serde::Serialize;

pub type ReceiveV1Response = Vec<ReceiveV1ResponseElement>;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReceiveV1ResponseElement {
    pub envelope: MessageEnvelope,
    pub account: String,
}

impl ReceiveV1ResponseElement {
    pub fn mock(message: &str, source_number: &str, account: &str) -> Self {
        let datamessage = DataMessage {
            timestamp: 0,
            message: message.to_string(),
            expires_in_seconds: 0,
            view_once: false,
            group_info: "".to_string(),
        };

        let message_envelope = MessageEnvelope {
            source: "".to_string(),
            source_number: source_number.to_string(),
            source_uuid: "".to_string(),
            source_name: "".to_string(),
            source_device: 0,
            timestamp: 0,
            data_message: datamessage,
        };

        ReceiveV1ResponseElement {
            envelope: message_envelope,
            account: account.to_string(),
        }
    }
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MessageEnvelope {
    pub source: String,
    pub source_number: String,
    pub source_uuid: String,
    pub source_name: String,
    pub source_device: u32,
    pub timestamp: u64,
    pub data_message: DataMessage,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DataMessage {
    pub timestamp: u64,
    pub message: String,
    pub expires_in_seconds: u64,
    pub view_once: bool,
    pub group_info: String,
}
